doc: nrf: app_dev: bootloaders_dfu: Add sysbuild image ID docs (#18152)
nrfconnect/sdk-nrf: doc/nrf/app_dev/bootloaders_dfu/sysbuild_image_ids.rst

| ``SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``     | Number of additional MCUboot images to include in MCUboot (see :ref:`sysbuild_assigned_images_ids_mcuboot_only`)                                     |
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES``       | If enabled, this option sets the number of MCUboot images for both MCUboot and the main application.                                                 |
|                                                        | If disabled, it sets the number only for the MCUboot image.                                                                                          |
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+

> 
> .. _sysbuild_assigned_images_ids_mcuboot_only:

Reviewer: ```suggestion

See :ref:`configuring_kconfig` for information about how to configure these options.

.. _sysbuild_assigned_images_ids_mcuboot_only:
```

Author: Fixed

MCUboot-only images
*******************

Some images are visible only to MCUboot, allowing it to manage them independently of the main application.
This setup supports features like shared secondary slots.
Currently, MCUboot uses this capability to update itself, which can be enabled with secure boot.
MCUboot then shares the main application's secondary slot for firmware updates.

> However, it directly transfers the update to the designated MCUboot image slot, which is then booted by |NSIB|.

Reviewer: ```suggestion
However, it directly transfers the update to the designated MCUboot image slot, which is then booted by :ref:`bootloader`.
```

Author: Fixed

The main application does not need to be aware of this process; it only needs to handle an MCUboot update as it would any normal application update and reboot to apply changes.
This functionality is possible because there are additional MCUboot-only images that are not accessible to the main application and are configured only within MCUboot.

The total number of images configured in MCUboot is the sum of ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES`` and ``SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``.
If ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES`` is enabled, the main application sets only ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES``.
Otherwise, it is not set.

Getting image IDs in an application
***********************************

When building with sysbuild, the image IDs are provided to the application through Kconfig values and can be used in the code as needed.

> These image IDs are also used in |NCS| hooks for tasks such as nRF5340 network core updates, so no additional configuration is required.

Reviewer: Is this an allusion to https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/device_guides/nrf53/serial_recovery.html ? Can we link there?
```suggestion
These image IDs are also used in |NCS| hooks for various tasks (such as :ref:`nRF5340 network core updates <ug_nrf5340_serial_recovery>`), so no additional configuration is required.
```
or maybe to https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/device_guides/nrf53/simultaneous_multi_image_dfu_nrf5340.html ? (In such case, the ref is `ug_nrf5340_multi_image_dfu`.)

Author: Fixed to second link

Image numbers
*************

Image numbers are assigned in ascending order based on the following priority:

> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
> | Image                | Value (if all enabled)         | Kconfig option (sysbuild)                       | Kconfig option (application/MCUboot)         |
> +======================+================================+=================================================+==============================================+
> | Application          | 0                              | ``SB_CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  | ``CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  |
> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
> | Network core         | 1                              | ``SB_CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` | ``CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` |
> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
> | nRF7x Wi-Fi patch    | 2                              | ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | ``CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` |
> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
> | QSPI XIP split image | 3                              | ``SB_CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER``     | ``CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER``     |
> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
> | MCUboot              | 4 (only set for MCUboot image) | ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      | ``CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      |
> +----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+

Reviewer 2: ```suggestion
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
| Image                | Value (if all enabled)         | Kconfig option (sysbuild)                       | Kconfig option (application/MCUboot)                         |
+======================+================================+=================================================+==============================================================+
| Application          | 0                              | ``SB_CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  | :kconfig:option:`CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER`    |
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
| Network core         | 1                              | ``SB_CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` | :kconfig:option:`CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`   |
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
| nRF7x Wi-Fi patch    | 2                              | ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | :kconfig:option:`CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`   |
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
| QSPI XIP split image | 3                              | ``SB_CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER``     | :kconfig:option:`CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER`       |
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
| MCUboot              | 4 (only set for MCUboot image) | ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      | :kconfig:option:`CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER`        |
+----------------------+--------------------------------+-------------------------------------------------+--------------------------------------------------------------+
```

Author: Fixed

MCUboot update package version
******************************

When MCUboot updates are enabled, the firmware embeds the |NSIB| version using the :kconfig:option:`CONFIG_FW_INFO_FIRMWARE_VERSION` Kconfig option.
However, during an update, MCUboot does not check this version; it only checks the MCUboot package version.

You can set the MCUboot update package version in sysbuild with ``SB_CONFIG_SECURE_BOOT_MCUBOOT_VERSION``.
You must increase this version number when deploying an MCUboot update.
If the version in the update is lower than the current version, MCUboot will reject the update and not transfer it to the opposing secure boot firmware slot.

Additionally, make sure to load the correct update image onto the device.

> If MCUboot is currently running from the s0 slot, then you must use the s1 update, and if it is running from the s1 slot, then you must use the s0 update.

Reviewer: ```suggestion
If MCUboot is currently running from the `s0` slot, then you must use the `s1` update, and if it is running from the `s1` slot, then you must use the `s0` update.
```

Author: Fixed, assuming these should have been \`\`

Reviewer: Yes, correct, ``.

.. _sysbuild_assigned_images_ids:

Sysbuild-assigned image IDs
###########################

.. contents::
   :local:
   :depth: 2

> When using sysbuild in |NCS| with MCUboot, IDs for images will be automatically assigned by the build system depending upon configuration of a project, this allows for features to be enabled and disabled and have the image IDs update automatically without needing manual changes to applications or the bootloader (code and/or configuration)

Author: Fixed

> The following configuration options are available in sysbuild which show the image IDs assigned to a particular image (a value of ``-1`` indicates the image is not present):
> 
> .. note::
>     These are also provided to MCUboot and the main application via Kconfig values too (with exception of ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER`` which is provided to MCUboot only)

Reviewer 2: ```suggestion
Sysbuild has the following configuration options that display the image IDs assigned to each image. 
A value of ``-1`` indicates that the image is not present:

.. note::
    These options are also shared with MCUboot and the main application through Kconfig values, except for the ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER`` option, which is exclusive to MCUboot.
```

Author: Fixed

| ``SB_CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  | :kconfig:option:`CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER`  | Image number for application update          | --                                                                                                           |
+-------------------------------------------------+------------------------------------------------------------+----------------------------------------------+--------------------------------------------------------------------------------------------------------------+
| ``SB_CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` | :kconfig:option:`CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER` | Image number for network core update         | nRF5340 device and ``SB_CONFIG_NETCORE_APP_UPDATE``                                                          |
+-------------------------------------------------+------------------------------------------------------------+----------------------------------------------+--------------------------------------------------------------------------------------------------------------+

> | ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | :kconfig:option:`CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER` | Image number for wifi-patch update           | nRF7x device used and ``SB_CONFIG_WIFI_PATCHES_EXT_FLASH_XIP`` or ``SB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE`` |

Reviewer 2: ```suggestion
| ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | :kconfig:option:`CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER` | Image number for Wi-Fi patch update           | nRF7x device used and ``SB_CONFIG_WIFI_PATCHES_EXT_FLASH_XIP`` or ``SB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE`` |
```

+-------------------------------------------------+------------------------------------------------------------+----------------------------------------------+--------------------------------------------------------------------------------------------------------------+
| ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      | :kconfig:option:`CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER`      | Image number for MCUboot update              | ``SB_CONFIG_SECURE_BOOT_APPCORE``                                                                            |
|                                                 | (only set for MCUboot image)                               |                                              |                                                                                                              |
+-------------------------------------------------+------------------------------------------------------------+----------------------------------------------+--------------------------------------------------------------------------------------------------------------+

> The following configuration options relate the to the number of images and how those image numbers are setup in images:

Reviewer 2: ```suggestion
The following configuration options specify the number of images and how these image numbers are configured:
```

Author: Fixed

> +--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------|
> | Kconfig option (sysbuild)                              | Description                                                                                                                                          |
> +========================================================+======================================================================================================================================================|

Reviewer 2: ```suggestion
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Kconfig option (sysbuild)                              | Description                                                                                                                                          |
+========================================================+======================================================================================================================================================+
```
fixes the render

| ``SB_CONFIG_MCUBOOT_MIN_UPDATEABLE_IMAGES``            | Minimum number of MCUboot images                                                                                                                     |
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+

> | ``SB_CONFIG_MCUBOOT_MIN_ADDITIONAL_UPDATEABLE_IMAGES`` | Minimum number of additional MCUboot images for MCUboot only (described below)                                                                       |

Reviewer 2: ```suggestion
| ``SB_CONFIG_MCUBOOT_MIN_ADDITIONAL_UPDATEABLE_IMAGES`` | Minimum number of additional MCUboot images for MCUboot only (see :ref:`sysbuild_assigned_images_ids_mcuboot_only`)                                  |
```

Author: Fixed

+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``SB_CONFIG_MCUBOOT_MAX_UPDATEABLE_IMAGES``            | Maximum number of MCUboot images                                                                                                                     |
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES``                | Number of MCUboot images to set in images                                                                                                            |
+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+

> | ``SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``     | Number of additional MCUboot images to include in MCUboot (described below)                                                                          |

Reviewer 2: ```suggestion
| ``SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``     | Number of additional MCUboot images to include in MCUboot (see :ref:`sysbuild_assigned_images_ids_mcuboot_only`)                                     |
```

Author: Fixed

+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+

> | ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES``       | If enabled then will set both MCUboot and the main application's number of MCUboot images, if not set then will only set this is the MCUboot image   |

Reviewer 2: ```suggestion
| ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES``       | If enabled, this option sets the number of MCUboot images for both MCUboot and the main application.                                                 |
|                                                        | If disabled, it sets the number only for the MCUboot image.                                                                                          |
```

Author: Fixed

+--------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------+

> MCUboot-only images

Reviewer 2: ```suggestion
.. _sysbuild_assigned_images_ids_mcuboot_only:

MCUboot-only images
```

Author: Fixed

*******************

> There are images that are only visible to MCUboot, this allows for MCUboot to deal with those images without interference/knowledge from the application, or to allow for features like shared secondary slots.
> At present, this is used by MCUboot to update MCUboot itself, which can enabled with secure boot, this will share the main application's secondary slot for storing firmware updates but will move the update directly to the opposing MCUboot image slot which the xNSIB?x can then boot (the application does not need to be aware of how this works, just that it can place an MCUboot update as it would a normal application update and reboot to apply).
> To allow for this, there are additional MCUboot-only images which are not presented to the main application and are only set in MCUboot.
> The total number of images set in MCUboot is ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES + SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``, whilst in the main application it is just ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES`` (if ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES`` is enabled, otherwise will not be set)

Reviewer 2: ```suggestion
Some images are visible only to MCUboot, allowing it to manage them independently of the main application.
This setup supports features like shared secondary slots. 
Currently, MCUboot uses this capability to update itself, which can be enabled with secure boot. 
MCUboot then shares the main application's secondary slot for firmware updates.
However, it directly transfers the update to the designated MCUboot image slot, which is then booted by |NSIB|.

The main application does not need to be aware of this process; it only needs to handle an MCUboot update as it would any normal application update and reboot to apply changes.
This functionality is possible because there are additional MCUboot-only images that are not accessible to the main application and are configured only within MCUboot. 

The total number of images configured in MCUboot is the sum of ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES`` and ``SB_CONFIG_MCUBOOT_ADDITIONAL_UPDATEABLE_IMAGES``.
If ``SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES`` is enabled, the main application sets only ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES``. 
Otherwise, it is not set.
```

Author: Fixed

Getting image IDs in an application
***********************************

> When building with sysbuild, the image IDs are provided to the application via Kconfig values, they can be used in the code as needed.
> These image IDs are also used in |NCS| hooks for things like nRF5340 network core updates, so no extra configuration is needed to configure these.

Reviewer 2: ```suggestion
When building with sysbuild, the image IDs are provided to the application through Kconfig values and can be used in the code as needed. 
These image IDs are also used in |NCS| hooks for tasks such as nRF5340 network core updates, so no additional configuration is required.
```

Image numbers
*************

> Image numbers will be assigned in ascending order with the following priority:

Reviewer 2: ```suggestion
Image numbers are assigned in ascending order based on the following priority:
```

Author: Fixed

+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
| Image                | Value (if all enabled)         | Kconfig option (sysbuild)                       | Kconfig option (application/MCUboot)         |
+======================+================================+=================================================+==============================================+
| Application          | 0                              | ``SB_CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  | ``CONFIG_MCUBOOT_APPLICATION_IMAGE_NUMBER``  |
+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
| Network core         | 1                              | ``SB_CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` | ``CONFIG_MCUBOOT_NETWORK_CORE_IMAGE_NUMBER`` |
+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+

> | nRF7x wifi patch     | 2                              | ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | ``CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` |

Reviewer 2: ```suggestion
| nRF7x Wi-Fi patch    | 2                              | ``SB_CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` | ``CONFIG_MCUBOOT_WIFI_PATCHES_IMAGE_NUMBER`` |
```

Author: Fixed

+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
| QSPI XIP split image | 3                              | ``SB_CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER``     | ``CONFIG_MCUBOOT_QSPI_XIP_IMAGE_NUMBER``     |
+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+
| MCUboot              | 4 (only set for MCUboot image) | ``SB_CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      | ``CONFIG_MCUBOOT_MCUBOOT_IMAGE_NUMBER``      |
+----------------------+--------------------------------+-------------------------------------------------+----------------------------------------------+

MCUboot update package version
******************************

> When MCUboot updates are supported, the MCUboot update will have the xNSIBx?x version embedded into the firmware using :kconfig:option:`CONFIG_FW_INFO_FIRMWARE_VERSION`, however this version is not checked by MCUboot when doing an update, only the MCUboot package version is checked.
> The MCUboot update package version is set in sysbuild with ``SB_CONFIG_SECURE_BOOT_MCUBOOT_VERSION``, it is important to up-issue this version number if deploying an MCUboot update as if the version of MCUboot in the update is lower, the update will be rejected by MCUboot and not placed into the opposing secure boot firmware slot.
> It is also important to load the correct update image to the device, if the device is currently running MCUboot from the s0 slot then the s1 update needs to be used, likewise if MCUboot is currently running from the s1 slot then the s0 update needs to be loaded, if the wrong slot image is uploaded, the update will be rejected by MCUboot.

Reviewer 2: ```suggestion
When MCUboot updates are enabled, the firmware embeds the |NSIB| version using the :kconfig:option:`CONFIG_FW_INFO_FIRMWARE_VERSION` Kconfig option. 
However, during an update, MCUboot does not check this version; it only checks the MCUboot package version.

You can set the MCUboot update package version in sysbuild with ``SB_CONFIG_SECURE_BOOT_MCUBOOT_VERSION``. 
You must increase this version number when deploying an MCUboot update. 
If the version in the update is lower than the current version, MCUboot will reject the update and not transfer it to the opposing secure boot firmware slot.

Additionally, make sure to load the correct update image onto the device. 
If MCUboot is currently running from the s0 slot, then you must use the s1 update, and if it is running from the s1 slot, then you must use the s0 update. 
If you upload to the wrong slot image, MCUboot will reject the update.
```

Author: Fixed
